File: src/main/java/com/soapdogg/lispInterpreter/printer/DotNotationPrinter.java
package com.soapdogg.lispInterpreter.printer;

import com.soapdogg.lispInterpreter.constants.ReservedValuesConstants;
import com.soapdogg.lispInterpreter.datamodels.AtomNode;
import com.soapdogg.lispInterpreter.datamodels.ExpressionNode;
import com.soapdogg.lispInterpreter.datamodels.Node;
import com.soapdogg.lispInterpreter.determiner.ExpressionNodeDeterminer;
import lombok.AllArgsConstructor;

import java.util.List;

@AllArgsConstructor(staticName = "newInstance")
public class DotNotationPrinter {

    private final ExpressionNodeDeterminer expressionNodeDeterminer;
    private final AtomNodePrinter atomNodePrinter;
    private final DotNotationExpressionNodePrinter dotNotationExpressionNodePrinter;

    public String printInDotNotation(List<Node> nodes) {
        if (nodes.isEmpty()) return ReservedValuesConstants.NIL + '\n';
        StringBuilder sb = new StringBuilder();
        for (Node node : nodes) {
            String dotNotation = printInDotNotation(node);
            sb.append(dotNotation);
            sb.append('\n');
        }
        return sb.toString();
    }

    public String printInDotNotation(Node node) {
        boolean isExpressionNode = expressionNodeDeterminer.isExpressionNode(node);
        if (isExpressionNode) return dotNotationExpressionNodePrinter.printExpressionNodeInDotNotation((ExpressionNode) node);
        return atomNodePrinter.printAtomNode((AtomNode)node);
    }
}

File: src/main/java/com/soapdogg/lispInterpreter/printer/DotNotationPrinter.kt
package com.soapdogg.lispInterpreter.printer

import com.soapdogg.lispInterpreter.constants.ReservedValuesConstants
import com.soapdogg.lispInterpreter.datamodels.*

class DotNotationPrinter (
    private val atomNodePrinter: AtomNodePrinter,
    private val dotNotationExpressionNodePrinter: DotNotationExpressionNodePrinter
) {

    fun printInDotNotation(nodes: List<NodeV2>): String {
        if (nodes.isEmpty()) return """${ReservedValuesConstants.NIL}${ReservedValuesConstants.NEW_LINE}"""
        return nodes.joinToString(
            ReservedValuesConstants.NEW_LINE.toString(),
            ReservedValuesConstants.EMPTY,
            ReservedValuesConstants.NEW_LINE.toString()
        ) {node -> printInDotNotation(node)}
    }

    fun printInDotNotation(node: NodeV2): String {
        return if (node is ExpressionListNode) dotNotationExpressionNodePrinter.printExpressionNodeInDotNotation(node)
        else atomNodePrinter.printAtomNode(node as AtomNode)
    }
}